### ISSUE TITLE
Consider making a `help` page for Reedline shortcuts like `!!`

### ISSUE THREAD

**Issue by webbedspace** [nushell/nushell#7698]
### Related problem

I'm somewhat familiar with shells, but I don't know every little trick, and I image most Nushell users are the same. For instance, I did not know about the `!!` shortcut (type `!!` then enter), nor that it already existed in bash and zsh, and only found out about it from the heavily outdated dataframe documentation page.

### Describe the solution you'd like

Therefore, I'd like it if there was a `help` page (a table similar to `help commands` et al) for these Reedline shortcuts, so that they could be discovered entirely within Nushell.

**Comment by fdncred** [comment]
Agreed. What do you think the command should be called?  In nushell `!!` and `!$` are called bashisms but i don't think `help bashisms` is a very discoverable name.

**Comment by webbedspace** [comment]
I think just "shortcuts" is probably the best that can be done here. Or maybe "expansions" in the sense that they expand to a full line (although people may confuse it with `~` etc.)

**Comment by fdncred** [comment]
Trying to think of everything we could put in this command.
|shortcut|description|example|
|-|-|-|
|!!|put the last command in the buffer|!!|
|!$|put the last argument from the last command in the buffer|!$|
|!\<idx from history -l\>|put the item from history idx number provided in the buffer|!5|
|!-\<number\>|put the nth last item in the buffer|!-5|
|~|user home path|cd ~|
|$nu| show nushell's special items|$nu|
|$env| show nushell's environment variables|$env|

**Comment by webbedspace** [comment]
Special variables and home path shouldn't be included IMO because they're a mostly different construct from the other four, which alter the buffer. There could be yet another help category for them too, I guess (comprising `$nu`, `$env`, `$in` and `$it`, and maybe `$nothing` until someone removes that 🐏)

**Comment by fdncred** [comment]
We already have `$nu.scope.vars` for variables but I think it would be nice to collect all this mess together in one place vs having separate commands or parameters.

TODO:
- `help constants` or `$nu.scope.constants`
- `help mut` or `$nu.scope.muts`

**Comment by amtoine** [comment]
@fdncred, i think your list is great, cannot think of anything missing right now :thinking: 

> We already have `$nu.scope.vars` for variables but I think it would be nice to collect all this mess together in one place vs having separate commands or parameters.
> 
> TODO:
> 
> * `help constants` or `$nu.scope.constants`
> * `help mut` or `$nu.scope.muts`
>
with the new `0.73` and `0.74`, these look almost mandatory to complete what `$nu.scope.vars` is already doing :yum: 

## about the name of the new command
we talked a bit about it in private with @fdncred and here's what i think
- `help shortcuts` is great for "~", "true", "false", etc
- could also be `help reserved`, 'cause they all look like special and reserved characters and strings for `nushell`
- what about `help built-in`? these are things built-in to `nushell`!
however, this might get confusing with commands in `help commands` whose `command_type` is `builtin` :thinking: 

for now that's all i've got for the names, right now i think `help shortcuts` is the best! :tada:

**Comment by amtoine** [comment]
> Special variables and home path shouldn't be included IMO because they're a mostly different construct from the other four, which alter the buffer. There could be yet another help category for the variables too, I guess (comprising `$nu`, `$env`, `$in` and `$it`, and maybe `$nothing` until someone removes that ram)
>
and now looking at this, i quite agree with that :thinking: 

maybe `$nu`, `$env`, `$in`, `$it` and `$nothing` should be in something like `help builtin` and the rest of the table above could go into something like `help shortcuts`

**Comment by fdncred** [comment]
ok. Here we go.
```nu
alias "help builtin-vars" = ([[var description]; ["$in" "the output of the previous pipeline"] ["$it" "an iterator"] ["$nu" "built in variables"] ["$env" "the evironment"] ["$nothing" "a nushell value representing null"]])
```
```
> help builtin-vars 
╭───┬──────────┬─────────────────────────────────────╮
│ # │   var    │             description             │
├───┼──────────┼─────────────────────────────────────┤
│ 0 │ $in      │ the output of the previous pipeline │
│ 1 │ $it      │ an iterator                         │
│ 2 │ $nu      │ built in variables                  │
│ 3 │ $env     │ the evironment                      │
│ 4 │ $nothing │ a nushell value representing null   │
╰───┴──────────┴─────────────────────────────────────╯
```

**Comment by amtoine** [comment]
> ok. Here we go.
> 
> ```nu
> alias "help builtin-vars" = ([[var description]; ["$in" "the output of the previous pipeline"] ["$it" "an iterator"] ["$nu" "built in variables"] ["$env" "the evironment"] ["$nothing" "a nushell value representing null"]])
> ```
> 
> ```
> > help builtin-vars 
> ╭───┬──────────┬─────────────────────────────────────╮
> │ # │   var    │             description             │
> ├───┼──────────┼─────────────────────────────────────┤
> │ 0 │ $in      │ the output of the previous pipeline │
> │ 1 │ $it      │ an iterator                         │
> │ 2 │ $nu      │ built in variables                  │
> │ 3 │ $env     │ the evironment                      │
> │ 4 │ $nothing │ a nushell value representing null   │
> ╰───┴──────────┴─────────────────────────────────────╯
> ```
>
that looks cool :+1: 

i think an "example" section would be beneficial, e.g. mentioning the following for `$in` and `$it` respectively
```bash
> def get-name [] { $in | get name }
> ls | get-name
```
and
```bash
ls | each {|it| print $it.name}
```

**Comment by webbedspace** [comment]
>`s | each {|it| print $it.name}`

That's not how `$it` works. `$it` refers to the following:
```
(ls).size | where $it > 5kb
```
It literally only works like this in `where` clauses (and some other places I can't remember).

**Comment by amtoine** [comment]
> That's not how `$it` works. `$it` refers to the following:
> 
> ```
> (ls).size | where $it > 5kb
> ```
> 
> It literally only works like this in `where` clauses (and some other places I can't remember).
> 
woopsie my bad, you are 100% right :relieved:

**Comment by fdncred** [comment]
IIRC, `$it` worked on all command that were "RowCondition" commands. We recently changed a bunch of those to work differently. Now, I'm thinking we only have 1 command like that now, which is `where`.

**Comment by nos1dot618** [comment]
Hey @amtoine, I'd like to work on this issue. I've been looking into it and would be happy to take it up. If it's still available, could you assign it to me?

**Comment by fdncred** [comment]
What'd you have in mind?

**Comment by nos1dot618** [comment]
> Trying to think of everything we could put in this command. I didn't even know the 3rd and 4th items were a thing until I looked at the code just now. LOL
> |shortcut|description|example|
> |-|-|-|
> |!!|put the last command in the buffer|!!|
> |!$|put the last argument from the last command in the buffer|!$|
> |!\<idx from history -l\>|put the item from history idx number provided in the buffer|!5|
> |!-\<number\>|put the nth last item in the buffer|!-5|
> |~|user home path|cd ~|
> |$nu| show nushell's special items|$nu|
> |$env| show nushell's environment variables|$env|
> |$in| whatever|$in|
> |$it| blah |$it|
> |$nothing| |$nothing|
> |null| |null|
> |true| |true|
> |false| |false|

Hello @fdncred, thank you for the response. I was planning to write a tutorial based on your above comment. I am already writing tutorials as part of this issue https://github.com/nushell/nushell/issues/18895, therefore I was thinking of taking this up in it.

**Comment by fdncred** [comment]
ok, sounds good. we'll see what you come up with.

**Comment by nos1dot618** [comment]
Hi @fdncred, I've raised a PR for this issue. It adds a new `tutor` reedline shortcuts tutorial and documents `$env` in the variable tutorial. Would appreciate your review when you get a chance!
